src/hooks/useForm.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useState, useMemo } from 'react';
+
+// Interfaces START
+
+interface Validator {
+  (value: string): string;
+}
+
+interface FormInput {
+  field: string;
+  value: string;
+  validators: Validator[];
+}
+interface ParameterFunction<T> {
+  (value: FormInput): T;
+}
+
+interface Dictionary<T> {
+  [key: string]: T;
+}
+
+// Interfaces END
+
+// Hook START
+const useForm = (settings: FormInput[]) => {
+  const createObjectFrom = (
+    arraySource: FormInput[],
+    keyFunction: ParameterFunction<string>,
+    valueFunction: ParameterFunction<any>
+  ) =>
+    arraySource.reduce((object: Dictionary<any>, item: FormInput) => {
+      object[keyFunction(item)] = valueFunction(item);
+      return object;
+    }, {});
+
+  const calculateErrorList = (
+    value: string,
+    errorFunctions: Validator[]
+  ): string[] => {
+    let errorArray: string[] = [];
+    for (let errorFunction of errorFunctions) {
+      let errorMessage: string = errorFunction(value);
+      if (errorMessage) {
+        errorArray.push(errorMessage);
+      }
+    }
+    return errorArray;
+  };
+
+  const errorFunctionsByKey: Dictionary<[]> = useMemo(() => {
+    return createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.validators || []
+    );
+  }, []);
+
+  const [fields, setFields] = useState<Dictionary<string>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.value || ''
+    )
+  );
+  const [errors, setErrors] = useState<Dictionary<[]>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => calculateErrorList(item.value, errorFunctionsByKey[item.field])
```

Comment:
If settings is of type interface FormInput[] with validators functions inside , why do we want to get those validators from here errorFunctionsByKey[item.field] ?
Can't we call calculateErrorList(item.value, item.validators) directly? Does it have to do with useMemo and performance? Please exaplain if it does :D

Reply:
It was a cold night, I remember very well that I heard a noise in the distance, the noise was getting closer and stronger, I was afraid. 

I took courage and decided to find out, apparently it was a mosquito.

Well leaving aside the good story that I will tell you later.


Yes, the reason is that I use a DRY pattern, I prefer to use this and leave the parameters as functions to be able to reuse the functionality

I decide to use useMemo to create an object only once, and in the future make big O (1) queries and access faster to the field validation functions

I also decide to pass functions to get the key, so thinking ahead I can create an object taking any value as key and any value as value of input parameter of the hook